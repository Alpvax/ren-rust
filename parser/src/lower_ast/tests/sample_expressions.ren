fun name => Str.join "\n" [ line (name ++ ", " ++ name ++ ", bo-") "b" name, line "banana-fana fo-" "f" name, line "fee-fi-mo-" "m" name, name ++ "!" ]

where _ is "a" => #true is "e" => #true is "i" => #true is "o" => #true is "u" => #true is _   => #false

where _ is "b" => #true is "f" => #true is "m" => #true is _   => #false

fun name => if is_vowel (Str.take 1 name) then name else Str.drop 1 name

fun prefix letter name => if letter == Str.take 1 name then prefix ++ shorten name ++ "\n" else prefix ++ letter ++ shorten name ++ "\n"

"string with ${"template \${sections}"}"
