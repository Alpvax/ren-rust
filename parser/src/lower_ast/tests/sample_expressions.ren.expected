[
    Lambda(
        [
            "PVar(\"name\")",
        ],
        Call(
            Scoped(
                [
                    "Str",
                ],
                "join",
            ),
            [
                Literal(
                    LStr(
                        [
                            Left(
                                "\\n",
                            ),
                        ],
                    ),
                ),
                Literal(
                    LArr(
                        [
                            Call(
                                Var(
                                    "line",
                                ),
                                [
                                    Binop(
                                        Var(
                                            "name",
                                        ),
                                        Concat,
                                        Binop(
                                            Literal(
                                                LStr(
                                                    [
                                                        Left(
                                                            ", ",
                                                        ),
                                                    ],
                                                ),
                                            ),
                                            Concat,
                                            Binop(
                                                Var(
                                                    "name",
                                                ),
                                                Concat,
                                                Literal(
                                                    LStr(
                                                        [
                                                            Left(
                                                                ", bo-",
                                                            ),
                                                        ],
                                                    ),
                                                ),
                                            ),
                                        ),
                                    ),
                                    Literal(
                                        LStr(
                                            [
                                                Left(
                                                    "b",
                                                ),
                                            ],
                                        ),
                                    ),
                                    Var(
                                        "name",
                                    ),
                                ],
                            ),
                            Call(
                                Var(
                                    "line",
                                ),
                                [
                                    Literal(
                                        LStr(
                                            [
                                                Left(
                                                    "banana-fana fo-",
                                                ),
                                            ],
                                        ),
                                    ),
                                    Literal(
                                        LStr(
                                            [
                                                Left(
                                                    "f",
                                                ),
                                            ],
                                        ),
                                    ),
                                    Var(
                                        "name",
                                    ),
                                ],
                            ),
                            Call(
                                Var(
                                    "line",
                                ),
                                [
                                    Literal(
                                        LStr(
                                            [
                                                Left(
                                                    "fee-fi-mo-",
                                                ),
                                            ],
                                        ),
                                    ),
                                    Literal(
                                        LStr(
                                            [
                                                Left(
                                                    "m",
                                                ),
                                            ],
                                        ),
                                    ),
                                    Var(
                                        "name",
                                    ),
                                ],
                            ),
                            Binop(
                                Var(
                                    "name",
                                ),
                                Concat,
                                Literal(
                                    LStr(
                                        [
                                            Left(
                                                "!",
                                            ),
                                        ],
                                    ),
                                ),
                            ),
                        ],
                    ),
                ),
            ],
        ),
    ),
    Lambda(
        [
            "PVar(\"name\")",
        ],
        If(
            Call(
                Var(
                    "is_vowel",
                ),
                [
                    Call(
                        Scoped(
                            [
                                "Str",
                            ],
                            "take",
                        ),
                        [
                            Literal(
                                LNum(
                                    1.0,
                                ),
                            ),
                            Var(
                                "name",
                            ),
                        ],
                    ),
                ],
            ),
            Var(
                "name",
            ),
            Call(
                Scoped(
                    [
                        "Str",
                    ],
                    "drop",
                ),
                [
                    Literal(
                        LNum(
                            1.0,
                        ),
                    ),
                    Var(
                        "name",
                    ),
                ],
            ),
        ),
    ),
    Lambda(
        [
            "PVar(\"prefix\")",
            "PVar(\"letter\")",
            "PVar(\"name\")",
        ],
        If(
            Binop(
                Var(
                    "letter",
                ),
                Eq,
                Call(
                    Scoped(
                        [
                            "Str",
                        ],
                        "take",
                    ),
                    [
                        Literal(
                            LNum(
                                1.0,
                            ),
                        ),
                        Var(
                            "name",
                        ),
                    ],
                ),
            ),
            Binop(
                Binop(
                    Var(
                        "prefix",
                    ),
                    Concat,
                    Call(
                        Var(
                            "shorten",
                        ),
                        [
                            Var(
                                "name",
                            ),
                        ],
                    ),
                ),
                Concat,
                Literal(
                    LStr(
                        [
                            Left(
                                "\\n",
                            ),
                        ],
                    ),
                ),
            ),
            Binop(
                Var(
                    "prefix",
                ),
                Concat,
                Binop(
                    Binop(
                        Var(
                            "letter",
                        ),
                        Concat,
                        Call(
                            Var(
                                "shorten",
                            ),
                            [
                                Var(
                                    "name",
                                ),
                            ],
                        ),
                    ),
                    Concat,
                    Literal(
                        LStr(
                            [
                                Left(
                                    "\\n",
                                ),
                            ],
                        ),
                    ),
                ),
            ),
        ),
    ),
]
